logic/transactions/pool.js

```diff
@@ -0,0 +1,922 @@
+
+		if (seconds > timeOut) {
+			__private.delete(transaction.id, poolList);
+			library.logger.info('Expired transaction: ' + transaction.id + ' received at: ' + transaction.receivedAt.toUTCString());
+			return setImmediate(eachSeriesCb);
+		} else {
+			return setImmediate(eachSeriesCb);
+		}
+	}, function (err) {
+		return setImmediate(cb, err, null);
+	});
+};
+
+/**
+ * Calculates timeout based on transaction.
+ * @private
+ * @param {transaction} transaction
+ * @return {number} timeOut
+ */
+__private.transactionTimeOut = function (transaction) {
+	if (transaction.type === transactionTypes.MULTI) {
+		return (transaction.asset.multisignature.lifetime * constants.secondsPerHour);
+	} else if (Array.isArray(transaction.signatures)) {
+		return (constants.unconfirmedTransactionTimeOut * constants.signatureTransactionTimeOutMultiplier);
+	} else {
+		return (constants.unconfirmedTransactionTimeOut);
+	}
+};
+
+/**
+ * Gets sender account, and verifies transaction.
+ * @private
+ * @implements {__private.setAccountAndGet}
+ * @implements {__private.verifyTransaction}
+ * @param {transaction} transaction
+ * @param {function} cb - Callback function
+ * @returns {setImmediateCallback} errors | sender
+ */
+__private.processUnverifiedTransaction = function (transaction, cb) {
+	async.waterfall([
+		async.apply(__private.setAccountAndGet, transaction),
+		__private.verifyTransaction
+	], function (err, transaction, sender) {
+		return setImmediate(cb, err, sender);
+	});
+};
+
+// Public methods
+/**
+ * Bounds input parameters to private variable modules.
+ * @param {Accounts} accounts
+ */
+TransactionPool.prototype.bind = function (accounts) {
+	modules = {
+		accounts: accounts,
+	};
+};
+
+/**
+ * Gets invalid, unverified, verified.pending and verified.ready counters.
+ * @implements {__private.countTransactionsPool}
+ * @return {Object} unverified, pending, ready
+ */
+TransactionPool.prototype.getUsage = function () {
+	return {
+		unverified: pool.unverified.count,
+		pending: pool.verified.pending.count,
+		ready: pool.verified.ready.count,
+		invalid: pool.invalid.count,
+		total: __private.countTransactionsPool()
+	};
+};
+
+/**
+ * Gets transaction based on transaction id.
+ * Checks all pool lists: unverified, pending, ready.
+ * @param {string} id
+ * @return {Object} transaction, status
+ */
+TransactionPool.prototype.get = function (id) {
+	var transaction = pool.unverified.transactions[id];
+	if (transaction !== undefined) {
+		return {
+			transaction: transaction,
+			status: 'unverified'
+		};
+	}
+
+	transaction = pool.verified.pending.transactions[id];
+	if (transaction !== undefined) {
+		return {
+			transaction: transaction,
+			status: 'pending'
+		};
+	}
+	
+	transaction = pool.verified.ready.transactions[id];
+	if (transaction !== undefined) {
+		return {
+			transaction: transaction,
+			status: 'ready'
+		};
+	}
+	return {
+		transaction: undefined,
```

Comment:
Why return `undefined`?

Reply:
Because transaction is not in the pool. Although it could be targeted as invalid transaction. But invalid queue main utility is to avoid multiple hits from hacked transactions.